### PR TITLE
Show Contact age on deceased date

### PR DESCRIPTION
**Pull request by sunilpawar**
Overview
----------------------------------------
If any contact is marked as deceased and deceased date is present, show the age of contact when it was deceased.

Before
----------------------------------------
We were not showing the age, we just show age for Normal contact.

**Deceased Contact with deceased date.**
age_before

**Normal contact with birth date and their age as per current date.**
age_normal_contact

After
----------------------------------------
**Show age of contact when it was deceased if deceased date is present.**
age_after

### CODE REVIEW

**Comment by yashodha** [comment]
@sunilpawar looks good. You might want to keep the display consistent as with alive contacts on the next line.

**Comment by sunilpawar** [comment]
@yashodha , lets wait for others input/suggestion, then i will proceed accordingly.

**Comment by colemanw** [comment]
I think I would word it like this:

_Date Deceased:_ **July 7th, 1991 (Age 30)**

File: CRM/Contact/BAO/Contact.php [hunk 1]
```diff
@@ -2599,7 +2599,11 @@ public static function getValues(&$params, &$values) {
       if ($contact->birth_date) {
         $birthDate = CRM_Utils_Date::customFormat($contact->birth_date, '%Y%m%d');
         if ($birthDate < date('Ymd')) {
-          $age = CRM_Utils_Date::calculateAge($birthDate);
+          $deceasedDate = NULL;
+          if (!empty($contact->is_deceased) && !empty($contact->deceased_date)) {
+            $deceasedDate = $contact->deceased_date;
+          }
+          $age = CRM_Utils_Date::calculateAge($birthDate, $deceasedDate);
```
Comment (eileenmcnaughton) [inline CRM/Contact/BAO/Contact.php:2606]: You could simplify this to 
```
CRM_Utils_Date::calculateAge($birthDate, $contact->deceased_date ?? NULL);
```

File: CRM/Utils/Date.php [hunk 2]
```diff
@@ -893,35 +893,41 @@ public static function getFromTo($relative, $from = NULL, $to = NULL, $fromTime
...
 
     $bDate = explode('-', $formatedBirthDate);
     $birthYear = $bDate[0];
     $birthMonth = $bDate[1];
     $birthDay = $bDate[2];
-    $year_diff = date("Y") - $birthYear;
+    if (empty($targetDate)) {
+      $targetDate = date('Y-m-d');
+    }
+    $targetDate = strtotime($targetDate);
```
Comment (eileenmcnaughton) [inline CRM/Utils/Date.php:913]: likewise this could be

```
$targetDate = strtotime($targetDate ?? 'now');
```

Comment (eileenmcnaughton) [reply CRM/Utils/Date.php:913]: or

```
$targetDate = strtotime($targetDate ?? date('Y-m'd');
```
if now is too time-inclusive

**Comment by eileenmcnaughton** [comment]
Seems like a good improvement - I think I agree with @colemanw's format

**Comment by sunilpawar** [comment]
updated the code as per @eileenmcnaughton  and @colemanw suggestion..

**Comment by colemanw** [comment]
This looks almost done. One last thing to fix: it appears to be showing age unconditionally but it would only be available if the contact has a deceased date AND a birth_date.

File: CRM/Contact/BAO/Contact.php (hunk 1 above)
Comment (sunilpawar) [reply CRM/Contact/BAO/Contact.php]: @eileenmcnaughton 
My changes were correct, if contact is marked `is_deceased` with `deceased_date`. Then unset the flag, `deceased_date` date remain present in DB even we unset the date in the form.

In this situation, we need to check both
`if (!empty($contact->is_deceased) && !empty($contact->deceased_date)) {`

Comment (sunilpawar) [reply CRM/Contact/BAO/Contact.php]:
```php
    [birth_date] => 1961-06-06
    [is_deceased] => 0
    [deceased_date] => 1991-07-07

```

**Comment by eileenmcnaughton** [comment]
@colemanw can you check the last revision & hopefully merge

**Comment by eileenmcnaughton** [comment]
@sunilpawar can you check that last thing?

**Comment by sunilpawar** [comment]
@eileenmcnaughton  i already made correction, only showing Age for deceased contact if both birth_date and deceased_date are present.

**Comment by eileenmcnaughton** [comment]
OK I tested and that last issue is addressed. This looks good, merging

**Comment by eileenmcnaughton** [comment]
erm - actually merging

**Pull request merged**